st2actions/st2actions/workflows/workflows.py

```diff
@@ -78,13 +94,61 @@ def process(self, message):
             raise ValueError(msg)
 
         try:
+            with self._semaphore:
+                self._active_messages += 1
             handler_function(message)
         except Exception as e:
             # If the exception is caused by DB connection error, then the following
             # error handling routine will fail as well because it will try to update
             # the database and fail the workflow execution gracefully. In this case,
             # the garbage collector will find and cancel these workflow executions.
             self.fail_workflow_execution(message, e)
+        finally:
+            with self._semaphore:
+                self._active_messages -= 1
+
+    def shutdown(self):
+        super(WorkflowExecutionHandler, self).shutdown()
+        while self._active_messages > 0:
+            concurrency.sleep(2)
```

Comment:
Should probably handle a timeout here to prevent an infinite while.

Reply:
Addressed